File: 760. Find Anagram Mappings/Solution.py
# https://leetcode.com/problems/find-anagram-mappings

class Solution:
	'''
    Time complexity: O(N), where n is the length of the input lists, because the dictionary d is created in O(n) time by iterating over the elements of nums2
                     The lookup operation in the dictionary d for each element in nums1 takes O(1) time, so the overall time complexity is O(n) for the entire code.

    Space complexity: O(N), because the dictionary d is created to store n key-value pairs and the output list also has n elements.
	'''
    def anagramMappings(self, nums1: List[int], nums2: List[int]) -> List[int]:
    # Create an empty dictionary called d
    d = {}
    
    # For each element j in nums2, create a key-value pair in dictionary d where the key is j and the value is its index
    for i, j in enumerate(nums2):
        d[j] = i
    
    # Create an empty list called res
    res = []
    
    # For each element i in nums1, append the value corresponding to the key i in dictionary d to the list res
    for i in nums1:
        res.append(d[i])
    
    # Return the list res
    return res

File: 760. Find Anagram Mappings/Solution.cpp
// https://leetcode.com/problems/find-anagram-mappings

class Solution {
public:
    /*
    Time complexity: O(N), where N is the size of the input vectors.
                     This is because the while loop that populates the map takes O(N) time and the
                     for loop that uses the map to construct the result vector also takes O(N) time.

    Space complexity: O(N), because the map and result vector each require O(N) space to store the input and output data.
    */
    vector<int> anagramMappings(vector<int>& nums1, vector<int>& nums2) {
        // Create a map that maps integer keys to integer values
        map<int, int> hm;
        // Create a vector to store the results
        vector<int> res;
        // Initialize a counter variable i to 0
        int i = 0;

        // Loop through all the elements in nums2
        while(i < nums2.size()) {
            // Add an entry to the map with the key as the element and the value as its index in nums2
            hm[nums2[i]] = i;
            i++; // Increment the counter variable
        }

        // Loop through all the elements in nums1
        for(int j=0; j < nums1.size(); j++){
            // Add to the results vector the value in the map corresponding to the key in nums1
            res.push_back(hm[nums1[j]]);
        }

        return res; // Return the results vector
    }
};